src/v/coproc/event_listener.cc

```diff
@@ -80,17 +88,69 @@ event_listener::persist_actions(absl::btree_map<script_id, iobuf> wsas) {
 }
 
 event_listener::event_listener(ss::sharded<pacemaker>& pacemaker)
-  : _client({config::shard_local_cfg().kafka_api()[0].address})
-  , _dispatcher(pacemaker, _abort_source) {}
+  : _dispatcher(pacemaker, _abort_source) {}
+
+ss::future<> event_listener::do_connect() {
+    vassert(!_client, "do_connect() called when theres already a valid client");
+    /// Invoke metadata request on a new client connected to this broker
+    auto client = std::make_unique<kafka::client::transport>(
+      rpc::base_transport::configuration{
+        .server_addr = ss::socket_address(
+          ss::net::ipv4_address("127.0.0.1"), 9092)});
+    co_await client->connect();
+    /// Parse the response looking for the broker that is the leader for the
+    /// model::coprocessor_internal_topic
+    kafka::metadata_response r = co_await client->dispatch(
+      kafka::metadata_request{.list_all_topics = true});
+    auto found = std::find_if(
+      r.topics.begin(), r.topics.end(), [](kafka::metadata_response::topic& t) {
+          return t.name == model::coprocessor_internal_topic;
+      });
+    if (found == r.topics.end()) {
+        co_return co_await ss::make_exception_future<>(
+          kafka::client::partition_error(
+            model::coprocessor_internal_tp,
+            kafka::error_code::unknown_topic_or_partition));
+    }
+    vassert(
+      found->partitions.size() == 1, "Copro internal topic misconfigured");
+    model::node_id id = found->partitions[0].leader;
+    auto broker = std::find_if(
+      r.brokers.begin(),
+      r.brokers.end(),
+      [id](const kafka::metadata_response::broker& b) {
+          return b.node_id == id;
+      });
+    /// If the broker is found set _client and connect (if neccessary)
+    vassert(broker != r.brokers.end(), "Erranous kafka metadata response");
+    ss::socket_address broker_addr(
+      ss::net::ipv4_address(broker->host), broker->port);
+    if (broker_addr == client->server_address()) {
+        _client = std::move(client);
+        co_return;
+    }
+    co_await client->stop();
+    _client = std::make_unique<kafka::client::transport>(
+      rpc::base_transport::configuration{.server_addr = broker_addr});
+    co_await _client->connect();
+}
 
 ss::future<> event_listener::start() {
-    return _client.connect().then([this] {
-        (void)ss::with_gate(_gate, [this] {
-            return ss::do_until(
-              [this] { return _abort_source.abort_requested(); },
-              [this] { return do_start(); });
-        });
+    (void)ss::with_gate(_gate, [this] {
+        return ss::do_until(
+          [this] { return _abort_source.abort_requested(); },
+          [this] {
+              if (!_client) {
+                  return do_connect().handle_exception(
```

Comment:
there is no reconnection handling here

Reply:
The loop will continue and we will call do_connect again because _client will be == nullptr